File: packages/js/src/paint-mixers.js
import { defaultModelArtifact } from './default-model-artifact.js';
import {
  defaultLearnedResidualCorrectionModel,
  LearnedResidualCorrectionModel,
} from './learned-residual-correction-model.js';
import { PipelinePaintMixer } from './pipeline-paint-mixer.js';
import {
  defaultSpectralBaseMixEngine,
  SpectralBaseMixEngine,
} from './spectral-base-mix-engine.js';

export const PaintMixers = {
  defaultArtifact() {
    return defaultModelArtifact;
  },

  default() {
    return new PipelinePaintMixer({
      baseEngine: defaultSpectralBaseMixEngine,
      correctionModel: defaultLearnedResidualCorrectionModel,
    });
  },

  spectralBase({ artifact = defaultModelArtifact } = {}) {
    if (artifact === defaultModelArtifact) {
      return defaultSpectralBaseMixEngine;
    }
    return new SpectralBaseMixEngine({ artifact });
  },

  learnedResidual({ artifact = defaultModelArtifact } = {}) {
    if (artifact === defaultModelArtifact) {
      return defaultLearnedResidualCorrectionModel;
    }
    return new LearnedResidualCorrectionModel({ artifact });
  },

  pipeline({ baseEngine, correctionModel = null, enforceCompatibility = true }) {
    return new PipelinePaintMixer({
      baseEngine,
      correctionModel,
      enforceCompatibility,
    });
  },
};

File: packages/kotlin/src/commonMain/kotlin/io/github/rtarik/paintmixer/PaintMixers.kt
package io.github.rtarik.paintmixer

object PaintMixers {
    private val spectralBaseEngine: BaseMixEngine = SpectralBaseMixEngine
    private val learnedResidualCorrection: ResidualCorrectionModel = LearnedResidualCorrectionModel
    private val defaultMixer: PaintMixer = PipelinePaintMixer(
        baseEngine = spectralBaseEngine,
        correctionModel = learnedResidualCorrection,
    )

    fun default(): PaintMixer = defaultMixer

    fun spectralBase(): BaseMixEngine = spectralBaseEngine

    fun learnedResidual(): ResidualCorrectionModel = learnedResidualCorrection

    fun pipeline(
        baseEngine: BaseMixEngine,
        correctionModel: ResidualCorrectionModel? = null,
    ): PaintMixer = PipelinePaintMixer(baseEngine = baseEngine, correctionModel = correctionModel)
}
